src/client/dfuse/ops/read.c

```diff
@@ -111,6 +111,333 @@ dfuse_readahead_reply(fuse_req_t req, size_t len, off_t position, struct dfuse_o
+			} else {
+				DFUSE_TRA_DEBUG(cd->ohs[i], "%#zx-%#zx read", position,
+						position + K128 - 1);
+				DFUSE_REPLY_BUFQ(cd->ohs[i], req, ev->de_iov.iov_buf + (i * K128),
+						 K128);
+			}
+
+			if (atomic_fetch_add_relaxed(&cd->exited, 1) == 7)
+				done = true;
+		}
+	} while (req && !done);
+
+	if (done) {
+		d_slab_release(cd->eqt->de_read_slab, cd->ev);
+		D_FREE(cd);
+	}
+}
+
+/* Submut a read to dfs.
+ *
+ * Returns true on success.
+ */
+static bool
+chunk_fetch(fuse_req_t req, struct dfuse_obj_hdl *oh, struct read_chunk_data *cd, int slot)
+{
+	struct dfuse_info        *dfuse_info = fuse_req_userdata(req);
+	struct dfuse_inode_entry *ie         = oh->doh_ie;
+	struct dfuse_event       *ev;
+	struct dfuse_eq          *eqt;
+	int                       rc;
+	daos_off_t                position = cd->bucket * CHUNK_SIZE;
+
+	eqt = pick_eqt(dfuse_info);
+
+	ev = d_slab_acquire(eqt->de_read_slab);
+	if (ev == NULL) {
+		cd->rc = ENOMEM;
+		return false;
+	}
+
+	ev->de_iov.iov_len = CHUNK_SIZE;
+	ev->de_req         = req;
+	ev->de_cd          = cd;
+	ev->de_sgl.sg_nr   = 1;
+	ev->de_len         = 0;
+	ev->de_complete_cb = chunk_cb;
+
+	cd->ev         = ev;
+	cd->eqt        = eqt;
+	cd->reqs[slot] = req;
+	cd->ohs[slot]  = oh;
+
+	rc = dfs_read(ie->ie_dfs->dfs_ns, ie->ie_obj, &ev->de_sgl, position, &ev->de_len,
+		      &ev->de_ev);
+	if (rc != 0)
+		goto err;
+
+	/* Send a message to the async thread to wake it up and poll for events */
+	sem_post(&eqt->de_sem);
+
+	/* Now ensure there are more descriptors for the next request */
+	d_slab_restock(eqt->de_read_slab);
+
+	return true;
+
+err:
+	daos_event_fini(&ev->de_ev);
+	d_slab_release(eqt->de_read_slab, ev);
+	cd->rc = rc;
+	return false;
+}
+
+/* Try and do a bulk read.
+ *
+ * Returns true if it was able to handle the read.
+ */
+static bool
+chunk_read(fuse_req_t req, size_t len, off_t position, struct dfuse_obj_hdl *oh)
+{
+	struct dfuse_inode_entry *ie = oh->doh_ie;
+	struct read_chunk_core   *cc;
+	struct read_chunk_data   *cd;
+	off_t                     last;
+	uint64_t                  bucket;
+	int                       slot;
+	bool                      submit = false;
+	bool                      rcb;
+
+	if (len != K128)
+		return false;
+
+	if ((position % K128) != 0)
+		return false;
+
+	last = D_ALIGNUP(position + len - 1, CHUNK_SIZE);
+
+	if (last > oh->doh_ie->ie_stat.st_size)
+		return false;
+
+	bucket = D_ALIGNUP(position + len, CHUNK_SIZE);
+	bucket = (bucket / CHUNK_SIZE) - 1;
+
+	slot = (position / K128) % 8;
+
+	DFUSE_TRA_DEBUG(oh, "read bucket %#zx-%#zx last %#zx size %#zx bucket %ld slot %d",
+			position, position + len - 1, last, ie->ie_stat.st_size, bucket, slot);
+
+	D_MUTEX_LOCK(&rc_lock);
+	if (ie->ie_chunk == NULL) {
+		D_ALLOC_PTR(ie->ie_chunk);
+		if (ie->ie_chunk == NULL)
+			goto err;
+		D_INIT_LIST_HEAD(&ie->ie_chunk->entries);
+	}
+	cc = ie->ie_chunk;
+
+	d_list_for_each_entry(cd, &cc->entries, list)
```

Comment:
Storing the buckets in a LRU list seems could have a high search cost when it will contains a lot of entries.
Maybe a structure such as a binary search tree could be more efficient ?

Reply:
It could - entry count is key here.  Entries are removed from the list when all slots in a bucket have been read so there should not really be more than one entry per client process outside of some very contrived I/O patterns and it'll only be that high if the app is using single-shard-file and clients are reading different offsets in it.

Reply:
I have to admit, that I am not familiar enough with HPC and AI use cases to know if this unfavorable use case is common.

Just to be sure to understand: if a client read do first aligned read of 128k, and never read the following 7 128K. Then the entry will stay in the cache as long as the client do not close the inode ?

Reply:
> Just to be sure to understand: if a client read do first aligned read of 128k, and never read the following 7 128K. Then the entry will stay in the cache as long as the client do not close the inode ?

Close - file handle rather than inode but yes.  The cache is tied to the inode but is dropped when the number of open file handles on that inode drops to zero.

Reply:
Thanks for the explanation.